Sources/HummingbirdCore/Server/HTTPServer.swift

```diff
@@ -22,162 +23,223 @@ import NIOTransportServices
-                    readTimeout: idleTimeoutConfiguration.readTimeout,
-                    writeTimeout: idleTimeoutConfiguration.writeTimeout
-                )
+        self.state = .initial(
+            responder: responder,
+            childChannelInitializer: childChannelInitializer,
+            onServerRunning: onServerRunning
+        )
+        self.additionalChannelHandlers = additionalChannelHandlers
+        self.logger = logger
+    }
+
+    /// Start server
+    /// - Parameter responder: Object that provides responses to requests sent to the server
+    /// - Returns: EventLoopFuture that is fulfilled when server has started
+    public func start() async throws {
+        switch self.state {
+        case .initial(let responder, let childChannelInitializer, let onServerRunning):
+            self.state = .starting
+            let (channel, quiescingHelper) = try await self.makeServer(
+                httpChannelInitializer: childChannelInitializer,
+                responder: responder
             )
+            // check state again
+            switch self.state {
+            case .initial, .running:
+                preconditionFailure("We should only be running once")
+
+            case .starting:
+                self.state = .running(channel: channel, quiescingHelper: quiescingHelper)
+                await onServerRunning()
+
+            case .shuttingDown, .shutdown:
+                try await channel.close()
+            }
+
+        case .starting, .running:
+            fatalError("Unexpected state")
+
+        case .shuttingDown:
+            throw Error.serverShuttingDown
+
+        case .shutdown:
+            throw Error.serverShutdown
         }
     }
 
-    /// Add TLS handler. Need to provide a closure so new instance of these handlers are
-    /// created for each child channel
-    /// - Parameters:
-    ///   - handler: autoclosure generating handler
-    ///   - position: position to place channel handler
-    @discardableResult public func addTLSChannelHandler(_ handler: @autoclosure @escaping () -> RemovableChannelHandler) -> Self {
-        self.tlsChannelHandler = handler
-        return self
+    /// Shutdown HTTP server
+    /// - Returns: EventLoopFuture that is fulfilled when server has stopped
+    public func shutdownGracefully() async throws {
+        switch self.state {
+        case .initial, .starting:
+            self.state = .shutdown
+
+        case .running(_, let quiescingHelper):
+            let promise = self.eventLoopGroup.next().makePromise(of: Void.self)
+            quiescingHelper.initiateShutdown(promise: promise)
+            self.state = .shuttingDown(shutdownPromise: promise)
+            try await promise.futureResult.get()
+
+            // We need to check the state here again since we just awaited above
+            switch self.state {
+            case .initial, .starting, .running, .shutdown:
+                fatalError("Unexpected state")
```

Comment:
I think that this fatalError could lead to really unfortunate crashes. Imagine a `deinit` or other tear-down that shuts down the server before it had a chance to boot. A unit test if you will, with a failing `setUp`. While an error is certainly expected and correct, I think that this is a bit of an overeager crash.

Reply:
Also, I might be wrong about this. But since we're checking the old state so regularly here, because of re-entrancy, I can't help but imagine there has to be a cleaner solution.

Reply:
This is only two places we double check the state. Every other check is from the root of a public function call. The only place you can exit the shuttingDown state is here. If the state has changed then there is a problem with the state machine. I guess in theory we could remove this check, it was just in there to be safe